proof-producer/libs/preset/include/nil/proof-generator/preset/preset.hpp

```diff
@@ -0,0 +1,53 @@
+#ifndef PROOF_GENERATOR_LIBS_PRESET_PRESET_HPP_
+#define PROOF_GENERATOR_LIBS_PRESET_PRESET_HPP_
+
+#include <boost/log/core.hpp>
+#include <boost/log/expressions.hpp>
+#include <boost/log/trivial.hpp>
+#include <istream>
+
+#include "nil/proof-generator/preset/bytecode.hpp"
+#include "nil/proof-generator/preset/add.hpp"
+
+#include <optional>
+#include <string>
+#include <unordered_map>
+
+namespace nil {
+    namespace proof_generator {
+        template<typename BlueprintFieldType>
+        class CircuitFactory {
+            static std::map<const std::string, std::function<std::optional<std::string>(
+                    std::optional<nil::crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType>>& circuit,
+                    std::optional<nil::crypto3::zk::snark::plonk_table<BlueprintFieldType, nil::crypto3::zk::snark::plonk_column<BlueprintFieldType>>>& assignment_table)>> circuit_selector;
+        public:
+            static std::optional<std::string> initialize_circuit(const std::string& circuit_name,
+                std::optional<nil::crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType>>& circuit,
+                std::optional<nil::crypto3::zk::snark::plonk_table<BlueprintFieldType, nil::crypto3::zk::snark::plonk_column<BlueprintFieldType>>>& assignment_table,
+                std::optional<nil::crypto3::zk::snark::plonk_table_description<BlueprintFieldType>>& desc) {
```

Comment:
If I understand correctly, desc here is an output argument, so why is it optional?

Reply:
because it `optionla` in `prover` class.
For send as reference need to initialize `table_description_` in `setup_prover` method, then send as reference and rewrite.
I guess better initialize it once.
The main responsibility of `setup_prover` method - initialize assignment_table, description and circuit